src/Storages/ObjectStorage/StorageObjectStorageSource.cpp

```diff
@@ -0,0 +1,624 @@
+    // When reading lots of tiny files, this prefetching almost doubles the throughput.
+    // For bigger files, parallel reading is more useful.
+    if (use_prefetch)
+    {
+        LOG_TRACE(log, "Downloading object of size {} with initial prefetch", object_size);
+
+        auto async_reader = object_storage->readObjects(
+            StoredObjects{StoredObject{key, /* local_path */ "", object_size}}, read_settings);
+
+        async_reader->setReadUntilEnd();
+        if (read_settings.remote_fs_prefetch)
+            async_reader->prefetch(DEFAULT_PREFETCH_PRIORITY);
+
+        return async_reader;
+    }
+    else
+    {
+        /// FIXME: this is inconsistent that readObject always reads synchronously ignoring read_method setting.
+        return object_storage->readObject(StoredObject(key), read_settings);
+    }
+}
+
+StorageObjectStorageSource::IIterator::IIterator(const std::string & logger_name_)
+    : logger(getLogger(logger_name_))
+{
+}
+
+StorageObjectStorage::ObjectInfoPtr StorageObjectStorageSource::IIterator::next(size_t processor)
+{
+    auto object_info = nextImpl(processor);
+
+    if (object_info)
+    {
+        LOG_TEST(logger, "Next key: {}", object_info->relative_path);
+    }
+
+    return object_info;
+}
+
+StorageObjectStorageSource::GlobIterator::GlobIterator(
+    ObjectStoragePtr object_storage_,
+    ConfigurationPtr configuration_,
+    const ActionsDAG::Node * predicate,
+    const NamesAndTypesList & virtual_columns_,
+    ContextPtr context_,
+    ObjectInfos * read_keys_,
+    size_t list_object_keys_size,
+    bool throw_on_zero_files_match_,
+    std::function<void(FileProgress)> file_progress_callback_)
+    : IIterator("GlobIterator")
+    , WithContext(context_)
+    , object_storage(object_storage_)
+    , configuration(configuration_)
+    , virtual_columns(virtual_columns_)
+    , throw_on_zero_files_match(throw_on_zero_files_match_)
+    , read_keys(read_keys_)
+    , file_progress_callback(file_progress_callback_)
+{
+    if (configuration->isNamespaceWithGlobs())
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Expression can not have wildcards inside namespace name");
+    }
+    else if (configuration->isPathWithGlobs())
+    {
+        const auto key_with_globs = configuration_->getPath();
+        const auto key_prefix = configuration->getPathWithoutGlobs();
+        object_storage_iterator = object_storage->iterate(key_prefix, list_object_keys_size);
+
+        matcher = std::make_unique<re2::RE2>(makeRegexpPatternFromGlobs(key_with_globs));
+        if (!matcher->ok())
+        {
+            throw Exception(
+                ErrorCodes::CANNOT_COMPILE_REGEXP,
+                "Cannot compile regex from glob ({}): {}", key_with_globs, matcher->error());
+        }
+
+        recursive = key_with_globs == "/**";
+        filter_dag = VirtualColumnUtils::createPathAndFileFilterDAG(predicate, virtual_columns);
+    }
+    else
```

Comment:
Do we really create `GlobIterator` when `configuration->isPathWithGlobs() == false` in some places?

Reply:
I guess here it is not always globs, but we do create glob iterator: https://github.com/ClickHouse/ClickHouse/blob/1601115ff099beb0af46136e2500c351b54a0e48/src/Storages/StorageS3Cluster.cpp#L91-L94

Reply:
I think I will remove this code and create a keys iterator if needed in *cluster code, so glob iterator will throw exception in ctor if passed path does not contain globs.